### PR TITLE
[optimization] Split csm ubo v3.6.0

### PR DESCRIPTION
**Pull request by troublemaker52025**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by troublemaker52025** [comment]
@cocos-robot run test cases

File: cocos/core/pipeline/define.ts [hunk 1]
```diff
@@ -296,18 +289,43 @@ export class UBOShadow {
...
-        new Uniform('cc_csmSplitsInfo', Type.FLOAT4, 1),
-        new Uniform('cc_csmInfo', Type.FLOAT4, 1),
     ], 1);
 }
 globalDescriptorSetLayout.layouts[UBOShadow.NAME] = UBOShadow.LAYOUT;
 globalDescriptorSetLayout.bindings[UBOShadow.BINDING] = UBOShadow.DESCRIPTOR;
 
+/**
+ * @en The uniform buffer object only for dir csm shadow(level: 1 ~ 4)
+ * @zh 阴影 UBO。
```
Comment (jk20012001) [inline cocos/core/pipeline/define.ts]: 级联阴影使用的UBO

File: native/cocos/renderer/pipeline/shadow/ShadowFlow.cpp [hunk 2]
```diff
@@ -77,6 +82,13 @@ void ShadowFlow::activate(RenderPipeline *pipeline) {
     const int32_t isLinear = 0;
     pipeline->setValue("CC_SHADOWMAP_USE_LINEAR_DEPTH", isLinear);
 
+    // 0: UNIFORM_VECTORS_LESS_EQUAL_64, 1: UNIFORM_VECTORS_GREATER_EQUAL_125.
+    const auto uniformVectors = std::min(pipeline->getDevice()->getCapabilities().maxVertexUniformVectors,
```
Comment (jk20012001) [inline native/cocos/renderer/pipeline/shadow/ShadowFlow.cpp]: check fragment uniform is enough

File: cocos/core/pipeline/shadow/shadow-flow.ts [hunk 3]
```diff
@@ -83,6 +87,12 @@ export class ShadowFlow extends RenderFlow {
         const isLinear = pipeline.device.gfxAPI === API.WEBGL ? 1 : 0;
         pipeline.macros.CC_SHADOWMAP_USE_LINEAR_DEPTH = isLinear;
 
+        // 0: UNIFORM_VECTORS_LESS_EQUAL_64, 1: UNIFORM_VECTORS_GREATER_EQUAL_125.
+        const uniformVectors = Math.min(pipeline.device.capabilities.maxVertexUniformVectors,
+            pipeline.device.capabilities.maxFragmentUniformVectors);
```
Comment (jk20012001) [inline cocos/core/pipeline/shadow/shadow-flow.ts]: check fragment uniform is enough

File: editor/assets/chunks/surfaces/includes/common-vs.chunk [hunk 4]
```diff
@@ -8,6 +8,9 @@
 // UBO
 #include <builtin/uniforms/cc-global>
 #include <builtin/uniforms/cc-shadow>
+#if CC_SUPPORT_CASCADED_SHADOW_MAP
```
Comment (jk20012001) [inline editor/assets/chunks/surfaces/includes/common-vs.chunk]: surface shader only use csm in FS

**Review by jk20012001 — APPROVED** [review]

**Comment by jk20012001** [comment]
https://github.com/cocos/engine-extensions/pull/132

File: cocos/core/pipeline/custom/ubos.ts [hunk 5]
```diff
@@ -474,9 +485,10 @@ export class PipelineUBO {
         if (mainLight && shadowFrameBufferMap.has(mainLight)) {
             ds.bindTexture(UNIFORM_SHADOWMAP_BINDING, shadowFrameBufferMap.get(mainLight)!.colorTextures[0]!);
         }
-        PipelineUBO.updateShadowUBOView(this._pipeline, this._shadowUBO, camera);
+        PipelineUBO.updateShadowUBOView(this._pipeline, this._shadowUBO, this._csmUBO, camera);
         ds.update();
         cmdBuffer.updateBuffer(ds.getBuffer(UBOShadow.BINDING), this._shadowUBO);
+        cmdBuffer[0].updateBuffer(ds.getBuffer(UBOCSM.BINDING), this._csmUBO);
```
Comment (dumganhar) [inline cocos/core/pipeline/custom/ubos.ts]: `const cmdBuffer = this._pipeline.device.commandBuffer;`
cmdBuffer[0] ? Is it an array?

File: native/cocos/renderer/pipeline/PipelineSceneData.h [hunk 6]
```diff
@@ -126,6 +128,7 @@ class CC_DLL PipelineSceneData : public RefCounted {
     CSMLayers *_csmLayers{nullptr};
     bool _isHDR{true};
     float _shadingScale{1.0F};
+    bool _isSupportCSM{true};
```
Comment (dumganhar) [inline native/cocos/renderer/pipeline/PipelineSceneData.h]: Put _isHDR after _isSupportCSM for better struct padding and save memory.

File: cocos/core/pipeline/custom/ubos.ts (hunk 5 above)
Comment (troublemaker52025) [reply cocos/core/pipeline/custom/ubos.ts]: it a mistake

File: native/cocos/renderer/pipeline/PipelineSceneData.h (hunk 6 above)
Comment (troublemaker52025) [reply native/cocos/renderer/pipeline/PipelineSceneData.h]: sure

**Comment by troublemaker52025** [comment]
@cocos-robot run test cases

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 7]
```diff
@@ -55,6 +55,13 @@ export class PipelineSceneData {
         this._shadingScale = val;
     }
 
+    public get isSupportCSM () {
```
Comment (pandamicro) [inline cocos/core/pipeline/pipeline-scene-data.ts]: `supportCSM` or `doSupportCSM` or `csmSupported`

File: native/cocos/renderer/pipeline/PipelineSceneData.h [hunk 8]
```diff
@@ -83,6 +83,8 @@ class CC_DLL PipelineSceneData : public RefCounted {
     inline void clearValidPunctualLights() { _validPunctualLights.clear(); }
     inline float getShadingScale() const { return _shadingScale; }
     inline void setShadingScale(float val) { _shadingScale = val; }
+    inline bool isSupportCSM() const { return _isSupportCSM; }
```
Comment (pandamicro) [inline native/cocos/renderer/pipeline/PipelineSceneData.h]: follow ts naming

**Review by pandamicro — CHANGES_REQUESTED** [review]

**Review by star-e — APPROVED** [review]

**Review by pandamicro — APPROVED** [review]

**Pull request merged**